### PR TITLE
Add documentation to many parts of the Rust codebase

### PR DESCRIPTION
**Pull request by Wuelle**
I added lots of documentation, mostly for `graphite-graphene::layers`. There is definitely still a lot of room for improvement but its better than nothing. I also added the `From<ShapeLayer>`, `From<TextLayer>` and `From<FolderLayer>` impls for `Layer` that i talked about in the discord.
 (Some functions are documented even though their name is self-explanatory - because it makes the documentation generated by `rustdoc` easier to read)

### CODE REVIEW

File: graphene/src/intersection.rs
Comment (Keavon) [inline graphene/src/intersection.rs:13]: "rectangle" would be clearer

File: graphene/src/error.rs
Comment (Keavon) [inline graphene/src/error.rs:4]: Nitpick: Graphene should be capitalized

File: graphene/src/intersection.rs
Comment (Wuelle) [reply graphene/src/intersection.rs:13]: hmmm is it a rectangle? i dont think a quad necessarily has 90° corners

Comment (Keavon) [reply graphene/src/intersection.rs:13]: Oh, you're right. My bad. Then I'd suggest calling it "quad" then since the full word, quadrilateral, is not very common. (It's also misspelled at the moment, but easier to just shorten it to its more common form.)

File: graphene/src/error.rs
Comment (Wuelle) [reply graphene/src/error.rs:4]: done.

**Review by TrueDoctor — COMMENTED** [review]
Really great progress so far!

File: graphene/src/intersection.rs
Comment (Wuelle) [reply graphene/src/intersection.rs:86]: I didnt want to get any more specific because there is no **guarantee** that in the future, it will always be the first point on the path, even though it is right now. Would
```
/// Returns a point on `path`.
/// This function will usually return the first point from the path's first segment,
/// but callers should not rely on this behaviour.
``` 
be better?

Comment (TrueDoctor) [reply graphene/src/intersection.rs:86]: Sounds good :+1:

**Comment by Wuelle** [comment]
Hm forgot to comment again so i'll do it now: I added all your suggestions, re-review please?

**Comment by TrueDoctor** [comment]
From a technical point of view this all looks good, will wait for @Keavon to give his approval after spell checking + linting the text

**Comment by Keavon** [comment]
> From a technical point of view this all looks good, will wait for @Keavon to give his approval after spell checking + linting the text

Sure, I can review it for that sort of thing pretty quickly if I don't have to actually think about the logic of it. Let me get started shortly.

**Review by TrueDoctor — APPROVED** [review]
We should probably reslove the few leftover todos at some point but I think it would make a lot of sense to merge this pr, thoughts @Keavon?

**Comment by Keavon** [comment]
My bad! I totally forgot about this. I was supposed to read through it, like, a month ago. I'm terribly sorry for dropping the ball on this! I'll get this done today.

**Comment by Wuelle** [comment]
All good, im currently pretty busy irl anyways. No need to hurry

**Comment by Keavon** [comment]
FINALLY done. So sorry for the wait. GDC then school starting got in the way of my last promised time to review it. But now I'm merging it. I guess I've found one of the types of things I am prone to procrastinate: code reviewing lots of non-code. Thank you for the contribution and patience!

**Pull request merged**
